source/iopipe/json/serialize.d

```diff
@@ -845,118 +836,511 @@ unittest
+    
+    if (str.length == 0)
+        return false;
+    
+    static bool initialized = false;
+    if (!initialized)
+    {
+        // These are expensive.
+        L = unicode.L;
+        Nd = unicode.Nd;
+        Nl = unicode.Nl;
+        Mn = unicode.Mn;
+        Mc = unicode.Mc;
+        Pc = unicode.Pc;
+        initialized = true;
+    }
+    
+    dchar firstChar = decodeFront!(UseReplacementDchar.yes)(str);
+    
+    if (firstChar !in L && firstChar != '$' && firstChar != '_')
+        return false;
+    
+    foreach (dchar c; str)
+    {
+        if (c !in L &&
+            c !in Nd &&
+            c !in Nl &&
+            c !in Mn &&
+            c !in Mc &&
+            c !in Pc &&
+            c != '$' &&
+            c != '_'
+        )
+            return false;
+    }
+    return true;
+}
+
+private void doIndentation(Char)(scope void delegate(const(Char)[]) w, ref FormattingOptions fo)
+{
+    if (fo.indentation)
+        foreach (i; 0 .. fo.indentationLevel)
+            w(fo.indentationText);
+}
+
+// Static Arrays
+void serializeImpl(T : V[n], Char, V, size_t n)(scope void delegate(const(Char)[]) w, ref T val, FormattingOptions fo = FormattingOptions.init)
 {
     auto arr = val;
-    serializeImpl(w, val[]);
+    serializeImpl(w, val[], fo);
 }
 
-void serializeImpl(T, Char)(scope void delegate(const(Char)[]) w, ref T val) if (is(T == enum))
+// Enums
+void serializeImpl(T, Char)(scope void delegate(const(Char)[]) w, auto ref T val, FormattingOptions fo = FormattingOptions.init)
+    if (is(T == enum))
 {
     // enums are special, serialize based on the name. Unless there's a UDA
     // saying to serialize as the base type.
     static if(hasUDA!(T, enumBaseType))
     {
-        serializeImpl(w, *cast(OriginalType!T*)&val);
+        serializeImpl(w, *cast(OriginalType!T*)&val, fo);
     }
     else
     {
-        import std.conv;
         auto enumName = val.to!string;
-        serializeImpl(w, enumName);
+        serializeImpl(w, enumName, fo);
     }
 }
 
-void serializeImpl(T, Char)(scope void delegate(const(Char)[]) w, ref T val) if (isDynamicArray!T && !isSomeString!T && !is(T == enum))
+// Arrays
+void serializeImpl(T : C[], Char, C)(scope void delegate(const(Char)[]) w, T val, FormattingOptions fo = FormattingOptions.init)
+    if (!isSomeString!T && !is(T == enum))
 {
-    // open brace
-    w("[");
+    if (val.length == 0)
+    {
+        w("[]");
+        return;
+    }
+    
+    if (fo.lineBreaks)
+        w("[\n");
+    else
+        w("[");
+    
+    FormattingOptions indented = fo.indent();
+    
+    doIndentation(w, indented);
+    
     bool first = true;
     foreach(ref item; val)
     {
-        if(first)
+        if (first)
             first = false;
         else
-            w(", ");
-        serializeImpl(w, item);
-    }
+        {
+            w(fo.spaceBetweenItems ? ", ": ",");
+            if (fo.lineBreaks)
+                w("\n");
+            doIndentation(w, indented);
```

Comment:
Shouldn't this only happen with line breaks?

Reply:
`doIndentation()` checks for `FormattingOptions.indentation` which is a bool property that checks for `lineBreaks` so it does work correctly here but I agree this is rather obfuscated for the reader.